ml-prims/src/linalg/reduce_rows_by_key.h

```diff
@@ -0,0 +1,339 @@
+
+        #ifdef RRBK_SHMEM
+        __shared__ KeyType sh_keys[RRBK_SHMEM_SZ];
+        #endif
+        int rows_per_partition = nrows/gridDim.z+1;
+        int start_row = blockIdx.z*rows_per_partition;
+        int end_row = start_row + rows_per_partition;
+        end_row = end_row > nrows ? nrows : end_row;
+
+        KeyType this_key = blockIdx.y;
+        if (this_key >= nkeys) return;
+        int this_col = threadIdx.x + blockIdx.x*blockDim.x;
+        if (this_col >= ncols) return;
+        
+        DataType sum = 0.0;
+
+        #ifdef RRBK_SHMEM
+        int sh_key_inx = 0;
+        #endif
+        for (int r = start_row; r < end_row; r++) {
+            #ifdef RRBK_SHMEM
+            if (0 == sh_key_inx%RRBK_SHMEM_SZ) {
+               for (int x=threadIdx.x;x<RRBK_SHMEM_SZ;x+=blockDim.x)
+                   sh_keys[x] = d_keys[r+x];
+               __syncthreads();
+            }
+            if (sh_keys[sh_key_inx] != this_key) continue; //No divergence since this_key is the
+                                                 // same for the whole block
+            sh_key_inx++;
+            #else
+            if (d_keys[r] != this_key) continue; //No divergence since this_key is the
+                                                 // same for the whole block
+            #endif
+            //if ((end_row-start_row) / (r-start_row) != this_key) continue;
+            sum += __ldcg(&d_A[r*lda+this_col]);
+        }
+        
+        if (sum != 0.0) myAtomicAdd(&d_sums[this_key*ncols+this_col], sum);
+}
+
+template <typename DataType, typename KeyType>
+void sum_rows_by_key_large_nkeys_rowmajor( const DataType *d_A, int lda, 
+                    KeyType *d_keys, int nrows, int ncols, int key_offset, 
+                    int nkeys, DataType *d_sums, cudaStream_t st)  {
+        // x-dim refers to the column in the input data
+        // y-dim refers to the key
+        // z-dim refers to a partitioning of the rows among the threadblocks
+        dim3 grid, block;
+        block.x = 256; //Adjust me!
+        block.y = 1; //Don't adjust me!
+        grid.x = (ncols+block.x-1)/block.x;
+        grid.y = nkeys;
+        grid.z = std::max(40960000/nkeys/ncols, (int)1); //Adjust me!
+        grid.z = std::min(grid.z, (unsigned int)nrows);
+        grid.z = std::min(grid.z, MAX_BLOCKS);
+        //std::cout << "block = " << block.x << ", " << block.y << std::endl;
+        //std::cout << "grid = " << grid.x << ", " << grid.y << ", " << grid.z << std::endl;
+        cudaMemset(d_sums, 0, sizeof(DataType)*nkeys*ncols);
+        sum_rows_by_key_large_nkeys_kernel_rowmajor<<<grid,block,0,st>>>(d_A, lda, d_keys, nrows, ncols, key_offset, nkeys, d_sums);   
+}
+
+
+
+/**
+ * @brief Computes the reduction of matrix rows for each given key 
+ * @tparam Greater whether to apply greater or lesser than comparison
```

Comment:
```suggestion
 * @tparam DayType the data type for d_A
 * @tparam KeyType the data type d_keys
```

Reply:
DayType => DataType.

I can't figure out how to make this change in the review.

Reply:
I corrected the typo in the review. Thanks.